### ISSUE TITLE
Frame history

### ISSUE THREAD

**Issue by bla-opfuckmorsy** [Pholey/Whiteboard#14]
tl;dr I'm trying to think up the cleanest way to store and manage previous images of the canvas so we only have to redraw past a certain point and won't have godaweful lag.

Since we now have more complex drawing tools, like wireframe and sketch, redrawing everything on the canvas becomes way more expensive as more actions are done.  Especially with the sketch tool; it doesn't generate much data, but the rendering algorithm is complex enough that rendering a bunch of them will eventually just freeze up and crash the program (n^2 time.. cbf to improve it).

So, obviously we need to have a way to only render stuff past a certain point.  For example, we might store a copy of the canvas image data for every 10 actions, so we only need to render at most 10 actions when we have to do a redraw.

The thing is, there are certain actions that will be very expensive to perform, for example a fill tool.  Earlier I tried to do a pixel-by-pixel eraser tool so we could have a round eraser brush, but I quickly found that even with fairly small amounts of image data, pixel-by-pixel is godaweful slow (at least, when you're doing if for every little mouse movement).

So, we can't do "every 10 actions", we have to have it smart, where we can tie frames to individual actions, and easily find out what frame we have to start at when redrawing, and we have to have a way of recognizing which brushes correspond to more expensive operations where we might need to store a frame every time an action is performed with that brush.

Also, for multi-user undo, we can have people removing actions in the middle of the action history, which means when we do a redraw, we have to stop and store new versions of frames for more recent parts of the action history.

I can probably implement this myself, by just mapping actions to frames somehow, but I thought I'd bring it up in case any of you have a really clean way to do this in mind.
